clap_derive/src/derives/subcommand.rs

```diff
@@ -291,3 +320,116 @@ fn gen_from_subcommand(
+        let sub_name = attrs.cased_name();
+        let variant_name = &variant.ident;
+        let (pattern, updater) = match variant.fields {
+            Named(ref fields) => {
+                let (fields, update): (Vec<_>, Vec<_>) = fields
+                    .named
+                    .iter()
+                    .map(|field| {
+                        let attrs = Attrs::from_field(
+                            field,
+                            parent_attribute.casing(),
+                            parent_attribute.env_casing(),
+                        );
+                        let field_name = field.ident.as_ref().unwrap();
+                        (
+                            quote!( ref mut #field_name ),
+                            from_arg_matches::gen_updater(&fields.named, &attrs, false),
+                        )
+                    })
+                    .unzip();
+                (quote!( { #( #fields, )* }), quote!( { #( #update )* } ))
+            }
+            Unit => (quote!(), quote!({})),
+            Unnamed(ref fields) => {
+                if fields.unnamed.len() == 1 {
+                    (
+                        quote!((ref mut arg)),
+                        quote!(::clap::FromArgMatches::update_from_arg_matches(
+                            arg, matches
+                        )),
+                    )
+                } else {
+                    abort_call_site!("{}: tuple enums are not supported", variant.ident)
+                }
+            }
+        };
+
+        quote! {
+            (#sub_name, #name :: #variant_name #pattern) => { #updater }
+        }
+    });
+
+    let child_subcommands = flatten_variants.iter().map(|(variant, attrs)| {
+        let sub_name = attrs.cased_name();
+        let variant_name = &variant.ident;
+        let (pattern, updater) = match variant.fields {
+            Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+                let ty = &fields.unnamed[0];
+                (
+                    quote!((ref mut arg)),
+                    quote! {
+                        <#ty as ::clap::Subcommand>::update_from_subcommand(arg, Some((name, matches)));
+                    },
+                )
+            }
+            _ => abort!(
+                variant,
+                "`flatten` is usable only with single-typed tuple variants"
+            ),
+        };
+        quote! {
+            (#sub_name, #name :: #variant_name #pattern) => { #updater }
+        }
+    });
+
+    quote! {
+        fn update_from_subcommand<'b>(
+            &mut self,
+            subcommand: Option<(&str, &::clap::ArgMatches)>
+        ) {
+            if let Some((name, matches)) = subcommand {
+                match (name, self) {
+                    #( #subcommands ),*
+                    #( #child_subcommands ),*
+                    (_, s) => if let Some(sub) = <Self as ::clap::Subcommand>::from_subcommand(Some((name, matches))) {
```

Comment:
I wonder if we don't need `update_from_subcommand` here.

Reply:
Could be discussed, it would require to add more logic to support updating external subcommands.

Reply:
Can you please create a detailed issue regarding this? I will merge this for now.

Reply:
https://github.com/clap-rs/clap/pull/1878#pullrequestreview-443198826 this is why I made it that way. Before the external subcommand would extend the argument vector instead of replacing it.

Reply:
Yeah, I am not exactly against this, but I just want to rethink this later in the future and also add a test. So, I was asking you to create a issue to track it.